src/conditions/datadog_search.rs

```diff
@@ -27,215 +26,297 @@ impl_generate_config_from_default!(DatadogSearchConfig);
+                })
+                .any(|value| value == field),
+            _ => false,
+        },
+        Field::Default(f) | Field::Facet(f) | Field::Reserved(f) => log.contains(&f),
     }
+}
 
-    fn prefix(&self, field: Field, prefix: &str) -> Box<dyn Matcher<LogEvent>> {
-        match field {
-            // Default fields are matched by word boundary.
-            Field::Default(field) => {
-                let re = word_regex(&format!("{}*", prefix));
+fn equals(field: &str, value: &str, log: &LogEvent) -> bool {
+    match log.get(field) {
+        Some(Value::Bytes(s)) => s == value.as_bytes(),
+        _ => false,
+    }
+}
 
-                string_match(field, move |value| re.is_match(&value))
+fn tag_exists(to_match: &str, log: &LogEvent) -> bool {
+    match log.get("tags") {
+        Some(Value::Array(values)) => values.iter().any(|value| {
+            if let Value::Bytes(bytes) = value {
+                bytes == to_match.as_bytes()
+            } else {
+                false
             }
-            // Tags are recursed until a match is found.
-            Field::Tag(tag) => {
-                let starts_with = format!("{}:{}", tag, prefix);
+        }),
+        _ => false,
+    }
+}
 
-                any_string_match("tags", move |value| value.starts_with(&starts_with))
+fn regex_match(field: &str, re: &Regex, log: &LogEvent) -> bool {
+    match log.get(field) {
+        Some(Value::Bytes(s)) => {
+            if let Ok(s) = std::str::from_utf8(s) {
+                re.is_match(s)
+            } else {
+                false
             }
-            // All other field types are compared by complete value.
-            Field::Reserved(field) | Field::Facet(field) => {
-                let prefix = prefix.to_owned();
+        }
+        _ => false,
+    }
+}
 
-                string_match(field, move |value| value.starts_with(&prefix))
+fn prefix(field: &Field, pfx: &str, log: &LogEvent) -> bool {
+    match field {
+        // Default fields are matched by word boundary.
+        Field::Default(field) => match log.get(field.as_str()) {
+            Some(Value::Bytes(v)) => {
+                let re = word_regex(&format!("{}*", pfx));
+                re.is_match(v)
             }
-        }
+            _ => false,
+        },
+        // Tags are recursed until a match is found.
+        Field::Tag(tag) => match log.get("tags") {
+            Some(Value::Array(values)) => {
+                let starts_with: String = format!("{}:{}", tag, pfx);
+                values
+                    .iter()
+                    .any(|val: &Value| val.coerce_to_bytes().starts_with(starts_with.as_bytes()))
+            }
+            _ => false,
+        },
+        // All other field types are compared by complete value.
+        Field::Reserved(field) | Field::Facet(field) => match log.get(field.as_str()) {
+            Some(Value::Bytes(v)) => v.starts_with(pfx.as_bytes()),
+            _ => false,
+        },
```

Comment:
Since we have the `Field` at build time, we should be able to pull the regex construction and `format!` up to the construction of the matcher. We did basically the same thing here: https://github.com/vectordotdev/vector/blob/a600d15c66af4d3317935b052d573ccc3d9a79f8/lib/datadog/filter/src/fast_matcher.rs#L99-L112

As written, I think this might actually be a regression from the previous behavior.

Reply:
That's an interesting call out. The results here were not dramatic. I'll correct this -- I buy your analysis -- and see if this actually is an improvement.

Reply:
I'm not sure that the prefix matching is hit in any of the soaks, so it may not move those numbers at all unfortunately.